Add cmake package config support

herbrechtsmeier opened the pull request:
Add support for CMake find_package command to msgpack. The package
exports the two targets msgpackc-static and msgpackc if shared is
enabled. The CMake find_package command works after installation
of msgpack and during build of another project which adds msgpack
as sub project via add_subdirectory command.

redboltz commented:
@herbrechtsmeier , thank you for sending the PR.
It seems that some errors are reported by travis-ci and appveyor. Could you check it?

By the way, `CMAKE_MINIMUM_REQUIRED (VERSION 2.8.6)` currently. 
https://github.com/msgpack/msgpack-c/blob/master/CMakeLists.txt#L1
Do you use only features that are supported by 2.8.6 ?

herbrechtsmeier commented:
I have pushed a new version.

The `CMAKE_MINIMUM_REQUIRED (VERSION 2.8.6) ` is a problem. The `TARGET_INCLUDE_DIRECTORIES` function was introduced with version 2.8.11 and the `EXPORT(EXPORT)` function with version 3.0. Is it possible to increase the CMake version or should I deactivate the package configuration generation for versions lower 3.0?

redboltz commented:
@nobu-k , you mentioned that some major distribution uses cmake 2.8.6 when we decided the version of cmake. I forgot which distribution. Could you check it?

redboltz commented:
I talked with @nobu-k, CentOS 6, the previous version support cmake 2.8. http://mirror.centos.org/centos/6/os/i386/Packages/cmake-2.8.12.2-4.el6.i686.rpm

@herbrechtsmeier , I'm not expert about cmake but I guess that cmake would support conditional branch by the version. Could you keep `CMAKE_MINIMUM_REQUIRED (VERSION 2.8.6)` , and checking the version the location you implement the package config support. If the version is 3.0 or later, the package config support is introduced, otherwise disabled. I think that it is better.

herbrechtsmeier commented:
@redboltz: Could we update to the CMake version of CentOS 6 (2.8.12)? This will support the package config support excluded `EXPORT(EXPORT)` which is only needed to 

> Create a file <filename> that may be included by outside projects to import targets from the current project’s build tree

redboltz commented:
Sounds good! Updating to 2.8.12 is ok.

herbrechtsmeier commented:
I have updated the pull request to use the project specific CMake style, increase the CMake version to 2.6.12 and add a CMake version guard around the `export(export)` function. Additionally I have added an alias for the msgpack target if `MSGPACK_ENABLE_SHARED` is not enabled and only use the msgpack headers in most tests.

redboltz commented:
@herbrechtsmeier , thank you for updating the PR.

> I have updated the pull request to use the project specific CMake style, increase the CMake version to 2.6.12 and add a CMake version guard around the export(export) function.

Do you mean 2.8.12? 

I'm not an expert of cmake, but the PR looks good to me. @nobu-k , if you have any comments, please post them. I will merge the PR on this weekend.

redboltz commented:
merged.
